polaris-core/src/main/java/org/apache/polaris/core/catalog/PageToken.java

```diff
@@ -0,0 +1,145 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.catalog;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Represents a page token that can be used by operations like `listTables`. Clients that specify a
+ * `pageSize` (or a `pageToken`) may receive a `next-page-token` in the response, the content of
+ * which is a serialized PageToken.
+ *
+ * <p>By providing that in the next query's `pageToken`, the client can resume listing where they
+ * left off. If the client provides a `pageToken` or `pageSize` but `next-page-token` is null in the
+ * response, that means there is no more data to read.
+ */
+public class PageToken {
+
+  public final int offset;
+  public final int pageSize;
+
+  private static final String TOKEN_PREFIX = "polaris";
+  private static final int TOKEN_START = 0;
+  private static final int DEFAULT_PAGE_SIZE = 1000;
+
+  public static PageToken DONE = null;
+
+  public PageToken(int offset, int pageSize) {
+    this.offset = offset;
+    this.pageSize = pageSize;
+  }
+
+  /** Construct a PageToken from a plain limit */
+  public static PageToken fromLimit(int limit) {
+    return new PageToken(TOKEN_START, limit);
+  }
+
+  /** Construct a PageToken to read everything */
+  public static PageToken readEverything() {
+    return new PageToken(TOKEN_START, Integer.MAX_VALUE);
+  }
+
+  /** Deserialize a token string into a PageToken object */
+  public static PageToken fromString(String tokenString) {
+    if (tokenString == null) {
+      return PageToken.readEverything();
+    } else if (tokenString.isEmpty()) {
+      return PageToken.fromLimit(DEFAULT_PAGE_SIZE);
+    } else {
+      try {
+        String decoded =
+            new String(Base64.getDecoder().decode(tokenString), StandardCharsets.UTF_8);
+        String[] parts = decoded.split(":");
+
+        if (parts.length != 4 || !parts[0].equals(TOKEN_PREFIX)) {
+          throw new IllegalArgumentException("Invalid token format in token: " + tokenString);
+        }
+
+        int offset = Integer.parseInt(parts[1]);
+        int pageSize = Integer.parseInt(parts[2]);
+        int checksum = Integer.parseInt(parts[3]);
+        PageToken token = new PageToken(offset, pageSize);
+
+        if (token.hashCode() != checksum) {
+          throw new IllegalArgumentException("Invalid checksum for token: " + tokenString);
+        } else {
+          return token;
+        }
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Failed to decode page token: " + tokenString, e);
+      }
+    }
+  }
+
+  /**
+   * Builds a new page token to reflect new data that's been read. If the amount of data read is
+   * less than the pageSize, this will return `PageToken.DONE` (done)
+   */
+  public PageToken updated(List<?> newData) {
+    if (newData == null || newData.isEmpty() || newData.size() < pageSize) {
```

Comment:
Just curious, should this be newData.size() <= pageSize? If so, we might not need newData.isEmpty(), assuming pageSize >=0.

Reply:
No, the amount of data in this page matching `pageSize` indicates that there could be more data to come. We would not want to return `PageToken.DONE` in that case.

Reply:
Assuming there 10 records, what happens in the following scenario?
page_size=3
3 + 3 + 3 + 1, what would the fourth request return?
page_size=5
5 + 5, what would the second request return?

Reply:
See the test -- in the fourth request in your scenario, the response would come back with no page token which the [spec](https://github.com/apache/iceberg/blob/ab2c6f889d07eeee51a1f58605be248e9330d91b/open-api/rest-catalog-open-api.yaml#L1697) describes as indicating there is no more data to be read:

> The server must return `null` value for the `next-page-token` in the last response.

In your second example, the second response would also come back with no page token.

Reply:
Assuming, you are referring to this test case. I think it covers my first scenario.
`      // List using the final token:
      PolarisPage<?> finalListResult = catalog.listViews(NS, secondListResult.pageToken);
      Assertions.assertThat(finalListResult.data.size()).isEqualTo(1);
      Assertions.assertThat(finalListResult.pageToken).isNull();
`

Can you try this also? I can also try out tomorrow...
`      // List with perfectly divisible pageSize
      PolarisPage<?> exactListResult = catalog.listNamespaces(PageToken.fromLimit(5));
      Assertions.assertThat(firstListResult.data.size()).isEqualTo(5);
      Assertions.assertThat(firstListResult.pageToken.toString()).isNotNull().isNotEmpty();
`

Reply:
Ah, I see. I have added a new test case [here](https://github.com/apache/polaris/pull/273/files#diff-465cb011ce85cca2fdfcfa100ca31ae8e4300734597ce87cfed6b240e63658a0R1344) to hopefully clarify this behavior.

It is true that a potentially wasteful listing operation could happen. So if there are 5000 entities and you use the default page size of 1000, then you will need to do **6** listing operations right now. Of course, this problem goes away if there are 4999 or 5001 entities. The reason is that we limit the metastore operation to retrieve only 1000 results, and so it has no way of knowing there wasn't actually any more data.

This will depend on the metastore implementation of course, but the current eclipselink metastore will exhibit this issue. One potential fix would be to increase the limit to e.g. pageSize + 1 and then to be a little smarter about the token generated for the response.

Do you think this is necessary?

Reply:
IMO, the bigger problem with the current implementation is that it is not really robust to any changes happening in the metastore. That is, if an entity is written while a paged list is happening, it's not clear how correct the results will be.

If we begin with entities [t1, t2, t3] and start a list operation with page size 1, it could be that we yield results like [t1, t1, t2, t3] if some t0 is inserted after the first page is retrieved.

Reply:
> pageSize + 1
This should help you eliminate an additional empty round trip from the client. You might have to check your logic once again to make sure there are no other edge cases.

Regarding the records changing between requests, I feel solving this will complicate your implementation. I don't know who the clients are going to be and how much tolerance they will have. E.g. if it is a compute client, then is it okay to skip one record? For UI clients, some edge cases might be tolerable.

Regarding embedding the entity ID, you might have to consider the sort key and order. Can the client give their own sort key and order? If so, your token needs to carry that forward between requests.

Sorry, I didn't mean to drag you into this complication :-) Depending upon your requirement, you could document the edge cases and limitations.

Reply:
Hi @dbosco, thanks for the reply. I think you framed the problem well and I have been reflecting on this. It seems to me that the requirement here needs to be that enable pagination will never cause results to be missing, and ideally will never cause duplicate results.

Accordingly I have refactored how the eclipse-link metastore implements the page tokens. 

As for your question about the extraneous listing, I think this is probably not something we need to optimize for. With the default page size being 1000, we can see that the number of listing events will increase by at most 0.1%.

Reply:
@eric-maynard, I quickly went through your code changes, and it involves quite a bit of refactoring :-) Overall, it looks good. All the best in getting this merged 👍